### PR TITLE
Enable configure-helper.sh to support two scenarios for etcd level encryption: decryption and adding encryption to existing clusters.

### PR DESCRIPTION
**Pull request by immutableT**
**What this PR does / why we need it**:

1. Enables two additional scenarios for etcd level encryption:
    Adding encryption to an existing cluster
   Decryption
Both of the above mentioned scenarios require that secrets be "touched" by the envelope transformer post cluster startup. This functionality is implemented in the apply-encryption-config function.
2. Refactor all encryption related logic into a single function - setup-etcd-encryption
3. Remove the responsibility of generating kms-plugin manifest from configure-helper.sh. Implementers of kms-plugins (cloud-providers) should rely on the up-stream components to generate kms-plugin manifest. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cluster/gce/gci/apiserver_manifest_test.go [hunk 1]
```diff
@@ -51,89 +55,138 @@ readonly DOCKER_REGISTRY="k8s.gcr.io"
 readonly ENABLE_LEGACY_ABAC=false
 readonly ETC_MANIFESTS=${KUBE_HOME}/etc/kubernetes/manifests
 readonly KUBE_API_SERVER_DOCKER_TAG=v1.11.0-alpha.0.1808_3c7452dc11645d-dirty
-readonly LOG_OWNER_USER=$(id -un)
+readonly LOG_OWNER_USER=$(whoami)
```
Comment (awly) [inline cluster/gce/gci/apiserver_manifest_test.go]: What's this change for?

File: cluster/gce/gci/apiserver_manifest_test.go [hunk 2]
```diff
@@ -51,89 +55,138 @@ readonly DOCKER_REGISTRY="k8s.gcr.io"
...
-readonly ETCD_KMS_KEY_ID={{.ETCDKMSKeyID}}
+{{if .CloudKMSIntegration}}
+readonly CLOUD_KMS_INTEGRATION=true
+{{end}}
 `
-	kubeAPIServerManifestFileName = "kube-apiserver.manifest"
-	kmsPluginManifestFileName     = "kms-plugin-container.manifest"
-	kubeAPIServerStartFuncName    = "start-kube-apiserver"
+	/*
+		   encryptionProviderConfigForKMS is base64 encoded yaml below
```
Comment (awly) [inline cluster/gce/gci/apiserver_manifest_test.go]: Could you base64-encode it at runtime? Makes it easier to edit.
The YAML itself can still be a string

File: cluster/gce/gci/apiserver_manifest_test.go [hunk 3]
```diff
@@ -51,89 +55,138 @@ readonly DOCKER_REGISTRY="k8s.gcr.io"
...
 type kubeAPIServerManifestTestCase struct {
 	*ManifestTestCase
-	apiServerContainer v1.Container
-	kmsPluginContainer v1.Container
 }
 
 func newKubeAPIServerManifestTestCase(t *testing.T) *kubeAPIServerManifestTestCase {
 	return &kubeAPIServerManifestTestCase{
-		ManifestTestCase: newManifestTestCase(t, kubeAPIServerManifestFileName, kubeAPIServerStartFuncName, []string{kmsPluginManifestFileName}),
+		ManifestTestCase: newManifestTestCase(t, kubeAPIServerManifestFileName, kubeAPIServerStartFuncName, []string{}),
```
Comment (awly) [inline cluster/gce/gci/apiserver_manifest_test.go]: `nil` instead of `[]string{}`

File: cluster/gce/gci/apiserver_manifest_test.go [hunk 4]
```diff
@@ -51,89 +55,138 @@ readonly DOCKER_REGISTRY="k8s.gcr.io"
...
+			}
+
+			c.invokeTest(e, deployHelperEnv)
+
+			execArgs := c.pod.Spec.Containers[0].Command[execArgsIndex]
+			flagIsInArg := strings.Contains(execArgs, encryptionConfigFlag)
+			flag := fmt.Sprintf("%s=%s", encryptionConfigFlag, e.EncryptionProviderConfigPath)
+
+			switch {
+			case tc.wantFlag && !flagIsInArg:
```
Comment (awly) [inline cluster/gce/gci/apiserver_manifest_test.go]: `case tc.wantFlag != flagIsInArg:`

File: cluster/gce/gci/apiserver_manifest_test.go [hunk 5]
```diff
@@ -51,89 +55,138 @@ readonly DOCKER_REGISTRY="k8s.gcr.io"
...
+				t.Fatalf("Got %q,\n want flags to contain %q", execArgs, flag)
+			case !tc.wantFlag && flagIsInArg:
+				t.Fatalf("Got %q,\n do not want flags to contain %q", execArgs, encryptionConfigFlag)
+			case tc.wantFlag && flagIsInArg:
+				if !strings.Contains(execArgs, flag) {
```
Comment (awly) [inline cluster/gce/gci/apiserver_manifest_test.go]: Move this check into the case condition above

File: cluster/gce/gci/configure-helper.sh [hunk 6]
```diff
@@ -1787,67 +1769,93 @@ EOM
...
+# 3. Add kms-socket-vol and kms-socket-vol-mnt to enable communication with kms-plugin (if requested)
+#
+# Expects parameters:
+# $1 - path to kube-apiserver template
+# $2 - kube-apiserver startup flags (must be passed by reference)
+#
+# Assumes vars (supplied via kube-env):
+# ENCRYPTION_PROVIDER_CONFIG
+# CLOUD_KMS_INTEGRATION
+
```
Comment (awly) [inline cluster/gce/gci/configure-helper.sh]: remove newline

File: cluster/gce/gci/configure-helper.sh [hunk 7]
```diff
@@ -1787,67 +1769,93 @@ EOM
...
     fi
 
-    # TODO: Validate that the encryption config is for KMS.
-
-    local kms_socket_dir="/var/run/kmsplugin"
-
-    # kms_socket_mnt is used by both kms_plugin and kube-apiserver - this is how these containers talk.
-    local kms_socket_mnt="{ \"name\": \"kmssocket\", \"mountPath\": \"${kms_socket_dir}\", \"readOnly\": false}"
+    local -n kube_api_server_params=$2
+    local encryption_provider_config_path=${ENCRYPTION_PROVIDER_CONFIG_PATH:-/etc/srv/kubernetes/encryption-provider-config.yml}
```
Comment (awly) [inline cluster/gce/gci/configure-helper.sh]: Mention `ENCRYPTION_PROVIDER_CONFIG_PATH` in the function comment above

File: cluster/gce/gci/configure-helper.sh [hunk 8]
```diff
@@ -1787,67 +1769,93 @@ EOM
...
+# Applies encryption provider config.
+# This function may be triggered in two scenarios:
+# 1. Decryption of etcd
+# 2. Encryption of etcd is added after the cluster is deployed
+# Both cases require that the existing secrets in etcd be re-proceeded.
+#
+# Assumes vars (supplied via kube-env):
+# ENCRYPTION_PROVIDER_CONFIG_FORCE
+function apply-encryption-config() {
+    if [[ -z "${ENCRYPTION_PROVIDER_CONFIG_FORCE:-}" ]] || [[ "${ENCRYPTION_PROVIDER_CONFIG_FORCE:-}" == "false" ]]; then
```
Comment (awly) [inline cluster/gce/gci/configure-helper.sh]: `if [[ "${ENCRYPTION_PROVIDER_CONFIG_FORCE:-false}" == "false" ]];`

File: cluster/gce/gci/configure-helper.sh [hunk 9]
```diff
@@ -1787,67 +1769,93 @@ EOM
...
+        return
+    fi
 
-      s@{{kms_socket_volume}}@${kms_socket_vol},@
-      s@{{encryption_provider_volume}}@${encryption_provider_vol},@
-    } " "${src_file}"
-  fi
+    # need kube-apiserver to be ready
+    until kubectl get secret
```
Comment (awly) [inline cluster/gce/gci/configure-helper.sh]: Hmm, what other startup steps are blocked by this? I wonder if this needs a timeout or maybe needs to run as the very last step of configure-helper

**Review by awly — COMMENTED** [review]
It's a lot of bash and sed,. Just to make sure have you successfully started a cluster with these changes?

**Comment by BenTheElder** [comment]
> It's a lot of bash and sed,. Just to make sure have you successfully started a cluster with these changes?

I'd love if you took a look at these with [shellcheck](https://github.com/koalaman/shellcheck), working on [a PR](https://github.com/kubernetes/kubernetes/pull/68438) to make this a presubmit currently.

File: cluster/gce/gci/configure-helper.sh (hunk 9 above)
Comment (immutableT) [reply cluster/gce/gci/configure-helper.sh]: If this block of code fails then we can safely assume that the cluster did not come-up, so I don't see much value in time-out.
I put this function into background to allow other things to continue.

File: cluster/gce/gci/apiserver_manifest_test.go (hunk 4 above)
Comment (immutableT) [reply cluster/gce/gci/apiserver_manifest_test.go]: This would save one case block, but will prevent me from providing a more informative error message.

**Comment by immutableT** [comment]
@awly PTAL.

File: cluster/gce/gci/configure-helper.sh (hunk 9 above)
Comment (awly) [reply cluster/gce/gci/configure-helper.sh]: Hmm, when the parent shell exits (when rest of configure-helper is done) will it continue running?
And will there be a way to tell if/when it fails?

**Comment by immutableT** [comment]
@BenTheElder ran shellcheck and addressed issues it found in my changes.

File: cluster/gce/gci/configure-helper.sh (hunk 9 above)
Comment (immutableT) [reply cluster/gce/gci/configure-helper.sh]: I believe that the parent will not exit until that function is done.
Today,  this is a weak spot - cluster creation times-out and users need to login to the master to investigate. In other words, the caller would not know why kube-up or cluster create fails.

**Comment by BenTheElder** [comment]
Thanks!!

File: cluster/gce/gci/configure-helper.sh [hunk 10]
```diff
@@ -1736,29 +1736,11 @@ function start-kube-apiserver {
...
-  fi
+  src_file="${src_dir}/kube-apiserver.manifest"
 
-  if [[ -n "${ENCRYPTION_PROVIDER_CONFIG:-}" ]]; then
-    ENCRYPTION_PROVIDER_CONFIG_PATH="${ENCRYPTION_PROVIDER_CONFIG_PATH:-/etc/srv/kubernetes/encryption-provider-config.yml}"
-    echo "${ENCRYPTION_PROVIDER_CONFIG}" | base64 --decode > "${ENCRYPTION_PROVIDER_CONFIG_PATH}"
-    params+=" --experimental-encryption-provider-config=${ENCRYPTION_PROVIDER_CONFIG_PATH}"
-  fi
+  # params is passed by reference, so no "$"
+  setup-etcd-encryption "${src_file}" params
```
Comment (mikedanese) [inline cluster/gce/gci/configure-helper.sh]: I'm confused. Isn't this params passed as a string?

File: cluster/gce/gci/configure-helper.sh [hunk 11]
```diff
@@ -1736,29 +1736,11 @@ function start-kube-apiserver {
...
-    - secrets
-    providers:
-    - kms:
-       name: grpc-kms-provider
-       cachesize: 1000
-       endpoint: unix:///var/run/kmsplugin/socket.sock
-EOM
-)
-  fi
+  src_file="${src_dir}/kube-apiserver.manifest"
```
Comment (mikedanese) [inline cluster/gce/gci/configure-helper.sh]: make local?

Comment (immutableT) [reply cluster/gce/gci/configure-helper.sh]: It is used in other places (ex. start-etcd-empty-dir-clean-up-pod).

Comment (mikedanese) [reply cluster/gce/gci/configure-helper.sh]: start-etcd-empty-dir-clean-up-pod declares src_file as local. Shouldn't we do the same here?

File: cluster/gce/gci/configure-helper.sh [hunk 12]
```diff
@@ -1787,67 +1769,105 @@ EOM
...
     fi
 
-    # TODO: Validate that the encryption config is for KMS.
-
-    local kms_socket_dir="/var/run/kmsplugin"
-
-    # kms_socket_mnt is used by both kms_plugin and kube-apiserver - this is how these containers talk.
-    local kms_socket_mnt="{ \"name\": \"kmssocket\", \"mountPath\": \"${kms_socket_dir}\", \"readOnly\": false}"
+    kube_api_server_params=$2
+    encryption_provider_config_path=${ENCRYPTION_PROVIDER_CONFIG_PATH:-/etc/srv/kubernetes/encryption-provider-config.yml}
```
Comment (mikedanese) [inline cluster/gce/gci/configure-helper.sh]: quote bash variable

File: cluster/gce/gci/configure-helper.sh [hunk 13]
```diff
@@ -1787,67 +1769,105 @@ EOM
...
+# Expects parameters:
+# $1 - path to kube-apiserver template
+# $2 - kube-apiserver startup flags (must be passed by reference)
+#
+# Assumes vars (supplied via kube-env):
+# ENCRYPTION_PROVIDER_CONFIG
+# CLOUD_KMS_INTEGRATION
+# ENCRYPTION_PROVIDER_CONFIG_PATH (will default to /etc/srv/kubernetes/encryption-provider-config.yml)
+function setup-etcd-encryption {
+    local kube_apiserver_template_path
```
Comment (mikedanese) [inline cluster/gce/gci/configure-helper.sh]: File uses two spaces for indent

File: cluster/gce/gci/configure-helper.sh [hunk 14]
```diff
@@ -1787,67 +1769,105 @@ EOM
...
+    # need kube-apiserver to be ready
+    until kubectl get secret
+    do
+        sleep 10
+    done
 
-  cp "${src_file}" "${ETC_MANIFESTS:-/etc/kubernetes/manifests}"
+    # forces all secrets to be re-written to etcd, and in the process either encrypting or decrypting them
+    # https://kubernetes.io/docs/tasks/administer-cluster/encrypt-data/
+    kubectl get secrets --all-namespaces -o json | kubectl replace -f -
```
Comment (mikedanese) [inline cluster/gce/gci/configure-helper.sh]: This seems like it will lose data if secrets are being modified while it runs.

**Comment by immutableT** [comment]
@mikedanese PTAL.

File: cluster/gce/gci/configure-helper.sh [hunk 15]
```diff
@@ -1822,67 +1804,118 @@ EOM
...
+function apply-encryption-config() {
+  if [[ "${ENCRYPTION_PROVIDER_CONFIG_FORCE:-false}" == "false" ]]; then
+      return
   fi
 
-  cp "${src_file}" "${ETC_MANIFESTS:-/etc/kubernetes/manifests}"
+  # need kube-apiserver to be ready
+  until kubectl get secret
+  do
+      sleep ${ENCRYPTION_PROVIDER_CONFIG_FORCE_DELAY:-5}
```
Comment (mikedanese) [inline cluster/gce/gci/configure-helper.sh]: fix indentation

File: cluster/gce/gci/configure-helper.sh [hunk 16]
```diff
@@ -1822,67 +1804,118 @@ EOM
...
+# This function may be triggered in two scenarios:
+# 1. Decryption of etcd
+# 2. Encryption of etcd is added after the cluster is deployed
+# Both cases require that the existing secrets in etcd be re-proceeded.
+#
+# Assumes vars (supplied via kube-env):
+# ENCRYPTION_PROVIDER_CONFIG_FORCE
+function apply-encryption-config() {
+  if [[ "${ENCRYPTION_PROVIDER_CONFIG_FORCE:-false}" == "false" ]]; then
+      return
```
Comment (mikedanese) [inline cluster/gce/gci/configure-helper.sh]: fix indentation

File: cluster/gce/gci/configure-helper.sh [hunk 17]
```diff
@@ -1822,67 +1804,118 @@ EOM
...
   fi
 
-  cp "${src_file}" "${ETC_MANIFESTS:-/etc/kubernetes/manifests}"
+  # need kube-apiserver to be ready
+  until kubectl get secret
+  do
```
Comment (mikedanese) [inline cluster/gce/gci/configure-helper.sh]: use single line until

https://google.github.io/styleguide/shell.xml?showone=Loops#Loops

File: cluster/gce/gci/configure-helper.sh [hunk 18]
```diff
@@ -1822,67 +1804,118 @@ EOM
...
+      sleep ${ENCRYPTION_PROVIDER_CONFIG_FORCE_DELAY:-5}
+  done
+
+  retries=${ENCRYPTION_PROVIDER_CONFIG_FORCE_RETRIES:-5}
+  # The command below may fail when a conflict is detected during an update on a secret (something
+  # else updated the secret in the middle of our update).
+  # TODO: Retry only on errors caused by a conflict.
+  until (( retries == 0 ))
+  do
```
Comment (mikedanese) [inline cluster/gce/gci/configure-helper.sh]: https://google.github.io/styleguide/shell.xml?showone=Loops#Loops

**Pull request merged**